File: CygniParser/src/cygni/lexer/Token.java
package cygni.lexer;

import java.util.HashMap;

public class Token {
    public int line;
    public int col;
    public Tag tag;
    public Object value;

    private static HashMap<String, Tag> keywords;

    static {
        keywords = new HashMap<String, Tag>();
        keywords.put("if", Tag.If);
        keywords.put("else", Tag.Else);
        keywords.put("var", Tag.Var);
        keywords.put("def", Tag.Def);
        keywords.put("return", Tag.Return);
        keywords.put("true", Tag.True);
        keywords.put("false", Tag.False);
        keywords.put("while", Tag.While);
        keywords.put("class", Tag.Class);
        keywords.put("private", Tag.Private);
    }

    public Token(int line, int col, Tag tag, Object value) {
        this.line = line;
        this.col = col;
        this.tag = tag;
        this.value = value;
        if (this.tag == Tag.Identifier) {
            String text = (String) this.value;
            if (keywords.containsKey(text)) {
                this.tag = keywords.get(text);
            }
        }
    }

    @Override
    public String toString() {
        if (value instanceof String) {
            return "t(Tag." + tag + ", \"" + value.toString() + "\")";
        } else {
            return "t(Tag." + tag + ", " + value.toString() + ")";
        }
    }

    @Override
    public boolean equals(Object o) {
        if (o instanceof Token) {
            Token t = (Token) o;
            return tag == t.tag & value.equals(t.value);
        } else {
            return false;
        }
    }
}

File: Cygni/src/Token.cpp
#include "Token.hpp"
#include <iostream>

namespace cygni
{
	Token::Token() :line{ -1 }, column{ -1 }, tag{ Tag::Eof }, text{}
	{
	}
	Token::Token(int line, int column, Tag tag, const std::u32string &text)
		: line{ line }, column{ column }, tag{ IdentifyKeyword(tag, text) }, text{ text } {}

	Tag Token::IdentifyKeyword(Tag tag, const std::u32string &text)
	{
		if (tag == Tag::Identifier)
		{
			if (keywords.find(text) != keywords.end())
			{
				return keywords[text];
			}
			else
			{
				return tag;
			}
		}
		else
		{
			return tag;
		}
	}

	std::unordered_map<std::u32string, Tag> Token::keywords = {
			{U"if", Tag::If},
			{U"else", Tag::Else},
			{U"var", Tag::Var},
			{U"def", Tag::Def},
			{U"return", Tag::Return},
			{U"true", Tag::True},
			{U"false", Tag::False},
			{U"void", Tag::Void},
			{U"while", Tag::While},
			{U"new", Tag::New},
			{U"class", Tag::Class},
			{U"private", Tag::Private},
			{U"module", Tag::Module},
			{U"protected", Tag::Protected},
			{U"package", Tag::Package},
			{U"import", Tag::Import},
			{U"require", Tag::Require},
			{U"rename", Tag::Rename},
			{U"to", Tag::To},
			{U"interface", Tag::Interface}
	};

	std::unordered_map<std::u32string, Tag> Token::operators = {
			{U"+", Tag::Add},
			{U"-", Tag::Subtract},
			{U"*", Tag::Multiply},
			{U"/", Tag::Divide},
			{U"%", Tag::Modulo},
			{U">", Tag::GreaterThan},
			{U"<", Tag::LessThan},
			{U">=", Tag::GreaterThanOrEqual},
			{U"<=", Tag::LessThanOrEqual},
			{U"==", Tag::Equal},
			{U"!=", Tag::NotEqual},
			{U"(", Tag::LeftParenthesis},
			{U")", Tag::RightParenthesis},
			{U"[", Tag::LeftBracket},
			{U"]", Tag::RightBracket},
			{U"{", Tag::LeftBrace},
			{U"}", Tag::RightBrace},
			{U":", Tag::Colon},
			{U",", Tag::Comma},
			{U".", Tag::Dot},
			{U";", Tag::Semicolon},
			{U"=", Tag::Assign},
			{U"->", Tag::RightArrow},
			{U"=>", Tag::GoesTo},
			{U"@", Tag::At},
			{U"<:", Tag::UpperBound},
			{U":>", Tag::LowerBound}
	};
} // namespace cygni
